### PR TITLE
And short descriptions to all solana-program modules and macros.

### PR DESCRIPTION
**Pull request by brson**
This patch cleans up the presentation of the solana-program crate API docs so that everything on the landing page has at least a description. After this patch these docs look quite nice.

In passing I also augmented or improved the docs in several modules, including `hash`, `decode_error`, `program_memory`, `sanitize`, `pubkey`.

As part of this work I filed one issue and one PR about API unsoundness:

- https://github.com/solana-labs/solana/issues/24035
- https://github.com/solana-labs/solana/pull/24091

### CODE REVIEW

File: sdk/program/src/sanitize.rs [hunk 1]
```diff
@@ -10,13 +12,15 @@ pub enum SanitizeError {
...
-/// signature verification checks, those are handled by another pass.
-/// Sanitize checks should include but are not limited too:
-///   * All index values are in range
-///   * All values are within their static max/min bounds
+/// A trait for sanitizing values and members of over the wire messages.
+///
+/// Implementation should recursively decent through the data structure and
+/// sanitize all struct members and enum clauses. Sanitize excludes signature
+/// verification checks, those are handled by another pass. Sanitize checks
+/// should include but are not limited to:
```
Comment (CriesofCarrots) [inline sdk/program/src/sanitize.rs:20]:
```suggestion
/// A trait for sanitizing values and members of over-the-wire messages.
///
/// Implementation should recursively descend through the data structure and
/// sanitize all struct members and enum clauses. Sanitize excludes signature-
/// verification checks, those are handled by another pass. Sanitize checks
/// should include but are not limited to:
```
(predates your changes)

File: sdk/program/src/pubkey.rs [hunk 2]
```diff
@@ -48,6 +50,20 @@ impl From<u64> for PubkeyError {
...
 
+/// The address of a [Solana account][acc].
+///
+/// Some account addresses are [ed25519] public keys, with corresponding secret
+/// keys that are managed off-chain. Often though account addresses do not have
+/// corresponding secret keys, as with [_program derived addresses_][pdas], or
+/// the secret key is not relevant to the operation of a program, and may have
+/// even been disposed of. As running Solana programs can not safely create or
+/// manage secret keys, the full [`Keypair`] is not defined in `solana-program`
+/// but in `solana-sdk`.
```
Comment (CriesofCarrots) [inline sdk/program/src/pubkey.rs:61]:
```suggestion
/// Some account addresses are [ed25519] public keys, with corresponding secret
/// keys that are managed off-chain. Often, though, account addresses do not have
/// corresponding secret keys &mdash; as with [_program derived addresses_][pdas] &mdash; or
/// the secret key is not relevant to the operation of a program, and may have
/// even been disposed of. As running Solana programs can not safely create or
/// manage secret keys, the full [`Keypair`] is not defined in `solana-program`
/// but in `solana-sdk`.
```

File: sdk/program/src/program_memory.rs [hunk 3]
```diff
@@ -1,10 +1,36 @@
-//! Solana Rust-based BPF memory operations
+//! Basic low-level memory operations.
+//!
+//! Within the BPF environment these are implemented as syscalls and executed by
```
Comment (CriesofCarrots) [inline sdk/program/src/program_memory.rs:3]:
```suggestion
//! Within the BPF environment, these are implemented as syscalls and executed by
```

File: sdk/program/src/program_memory.rs [hunk 4]
```diff
@@ -1,10 +1,36 @@
...
+/// # Arguments
+///
+/// - `dst` - Destination
+/// - `src` - Source
+/// - `n` - Number of bytes to copy
+///
+/// # Errors
+///
+/// When executed within a BPF program, the memory regions spanning `n` bytes
+/// from from the start of `dst` and `src` must be mapped program memory. If not
```
Comment (CriesofCarrots) [inline sdk/program/src/program_memory.rs:17]:
```suggestion
/// from from the start of `dst` and `src` must be mapped program memory. If not,
```

File: sdk/program/src/program_memory.rs [hunk 5]
```diff
@@ -1,10 +1,36 @@
...
+/// the program will abort.
+///
+/// The memory regions spanning `n` bytes from `dst` and `src` from the start
+/// of`dst` and `src` must not overlap. If they do then the program will abort
```
Comment (CriesofCarrots) [inline sdk/program/src/program_memory.rs:21]:
```suggestion
/// of `dst` and `src` must not overlap. If they do, then the program will abort
```

File: sdk/program/src/program_memory.rs [hunk 6]
```diff
@@ -21,13 +47,25 @@ pub fn sol_memcpy(dst: &mut [u8], src: &[u8], n: usize) {
...
+/// - `src` - Source
+/// - `n` - Number of bytes to copy
+///
+/// # Errors
 ///
-/// @param dst - Destination
-/// @param src - Source
-/// @param n - Number of bytes to copy
+/// When executed within a BPF program, the memory regions spanning `n` bytes
+/// from from `dst` and `src` must be mapped program memory. If not the program
```
Comment (CriesofCarrots) [inline sdk/program/src/program_memory.rs:61]:
```suggestion
/// from from `dst` and `src` must be mapped program memory. If not, the program
```

File: sdk/program/src/program_memory.rs [hunk 7]
```diff
@@ -42,11 +80,30 @@ pub unsafe fn sol_memmove(dst: *mut u8, src: *mut u8, n: usize) {
...
-/// @param s2 - Slice to be compared
-/// @param n - Number of bytes to compare
+/// - `s1` - Slice to be compared
+/// - `s2` - Slice to be compared
+/// - `n` - Number of bytes to compare
+///
+/// # Errors
+///
+/// When executed within a BPF program, the memory regions spanning `n` bytes
+/// from from the start of `dst` and `src` must be mapped program memory. If not
```
Comment (CriesofCarrots) [inline sdk/program/src/program_memory.rs:94]:
```suggestion
/// from from the start of `dst` and `src` must be mapped program memory. If not,
```

File: sdk/program/src/program_memory.rs [hunk 8]
```diff
@@ -67,11 +124,30 @@ pub fn sol_memcmp(s1: &[u8], s2: &[u8], n: usize) -> i32 {
...
+/// # Arguments
+///
+/// - `s` - Slice to be written
+/// - `c` - Byte to write
+/// - `n` - Number of bytes to write
+///
+/// # Errors
+///
+/// When executed within a BPF program, the memory region spanning `n` bytes
+/// from from the start of `s` must be mapped program memory. If not the program
```
Comment (CriesofCarrots) [inline sdk/program/src/program_memory.rs:138]:
```suggestion
/// from from the start of `s` must be mapped program memory. If not, the program
```

**Review by CriesofCarrots — COMMENTED** [review]
This is great!
One question in the `hash.rs` module, and a handful of nits.
@jackcmay , can you please look over the `program_memory.rs` docs?

File: sdk/program/src/program_memory.rs [hunk 9]
```diff
@@ -1,10 +1,36 @@
...
+/// from from the start of `dst` and `src` must be mapped program memory. If not
+/// the program will abort.
+///
+/// The memory regions spanning `n` bytes from `dst` and `src` from the start
+/// of`dst` and `src` must not overlap. If they do then the program will abort
+/// or, if run outside of the BPF VM, will panic.
+///
+/// # Safety
+///
+/// __This function is incorrectly missing an `unsafe` declaration.__
```
Comment (jackcmay) [inline sdk/program/src/program_memory.rs:26]: How about add `unsafe` to these?

File: sdk/program/src/hash.rs [hunk 10]
```diff
@@ -8,10 +10,20 @@ use {
     thiserror::Error,
 };
 
+/// Size of a hash in bytes.
 pub const HASH_BYTES: usize = 32;
-/// Maximum string length of a base58 encoded hash
+/// Maximum string length of a base58 encoded hash.
 const MAX_BASE58_LEN: usize = 44;
 
+/// A SHA-256 hash.
```
[the remark this answers is not in the document]
Comment (brson) [reply sdk/program/src/hash.rs]: I have made the suggested change, and updated the crate docs to reflect it.

**Comment by brson** [comment]
Thanks for the great reviews. I have made all suggested changes, except that I am going to make the `program_memory` breaking API changes in a different PR that I will post shortly. It should probably be merged prior to this one.

**Comment by brson** [comment]
I opened another PR for the API changes https://github.com/solana-labs/solana/pull/24362

**Review by CriesofCarrots — APPROVED** [review]
Lgtm, pending however you and @jackcmay decide to handle the mem op functions.

**Comment by brson** [comment]
CI error is probably not related to this patch.

**Pull request merged**